soc/xtensa/intel_adsp/ace/power.c

```diff
@@ -381,4 +381,28 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 	z_xt_ints_on(core_desc[cpu].intenable);
 }
 
+#ifdef CONFIG_ARCH_CPU_IDLE_CUSTOM
+
+#include <pm_data.h>
+
+void arch_cpu_idle(void)
+{
+#if CONFIG_ADSP_DYNAMIC_CLOCK_SWITCHING
+	extern void adsp_clock_idle_entry(void);
+	extern void adsp_clock_idle_exit(void);
+	struct intel_adsp_ace_pm_data *data =
+		(struct intel_adsp_ace_pm_data *) pm_state_custom_data_get(PM_STATE_ACTIVE, 0);
+
+	if (data && decrement_clock_switch_lock(data))
```

Comment:
This check is not enough to protect this region .. another core can successfully ` increment_clock_switch_lock(data)` and call `adsp_clock_idle_exit();` after this check and before this core executes `adsp_clock_idle_entry();` You need probably a spin lock around these operations.

Reply:
Agreed, this looks racy.

Reply:
It seems I didn't think it through carefully enough. Thanks.